### PR TITLE
Fixed NMSBoxes bug in DetectionModel

### PR DESCRIPTION
**Pull request by l-bat**
### Pull Request Readiness Checklist

related: opencv/opencv#17332

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under OpenCV (BSD) License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Custom,Custom Win,Custom Mac
build_image:Custom=ubuntu-openvino-2020.2.0:16.04
build_image:Custom Win=openvino-2020.2.0
build_image:Custom Mac=openvino-2020.2.0

test_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

File: modules/dnn/src/model.cpp [hunk 1]
```diff
@@ -246,6 +246,20 @@ void DetectionModel::detect(InputArray frame, CV_OUT std::vector<int>& classIds,
...
+    if (lastLayer->type == "Region")
+    {
+        for (String& name : impl->outNames)
+        {
+            int layerId = getLayerId(name);
+            Ptr<RegionLayer> layer = getLayer(layerId).dynamicCast<RegionLayer>();
+            CV_Assert(!layer.empty());
+            layer->nmsThreshold = 0;
+        }
+    }
```
Comment (dkurt) [inline modules/dnn/src/model.cpp:262]: Can we do it once? In some `init` method called from constructors?

File: modules/dnn/src/model.cpp [hunk 2]
```diff
@@ -347,20 +357,33 @@ void DetectionModel::detect(InputArray frame, CV_OUT std::vector<int>& classIds,
     else
         CV_Error(Error::StsNotImplemented, "Unknown output layer type: \"" + lastLayer->type + "\"");
 
-    if (nmsThreshold)
+    if (nmsThreshold && lastLayer->type == "Region")
```
Comment (dkurt) [inline modules/dnn/src/model.cpp:360]: Can we move it to `else if (lastLayer->type == "Region")`?

File: modules/dnn/src/model.cpp [hunk 3]
```diff
@@ -347,20 +357,33 @@ void DetectionModel::detect(InputArray frame, CV_OUT std::vector<int>& classIds,
...
+            NMSBoxes(localBoxes, localConfidences, confThreshold, nmsThreshold, indices);
+            for (int idx : indices)
+            {
+                boxes.push_back(localBoxes[idx]);
+                confidences.push_back(localConfidences[idx]);
+                classIds.push_back(it->first);
+            }
         }
     }
     else
```
Comment (dkurt) [inline modules/dnn/src/model.cpp:389]: Please remove 3 empty lines: https://github.com/opencv/opencv/blob/50d908c92bbbcc3b80d06276979c1ebf1898f923/modules/dnn/src/model.cpp#L395-L397

File: modules/dnn/include/opencv2/dnn/dnn.hpp [hunk 4]
```diff
@@ -1279,6 +1279,8 @@ CV__DNN_INLINE_NS_BEGIN
          CV_WRAP void detect(InputArray frame, CV_OUT std::vector<int>& classIds,
                              CV_OUT std::vector<float>& confidences, CV_OUT std::vector<Rect>& boxes,
                              float confThreshold = 0.5f, float nmsThreshold = 0.0f);
+        private:
+            void init();
```
Comment (dkurt) [inline modules/dnn/include/opencv2/dnn/dnn.hpp:1283]: Can we avoid adding it to a public header?

File: modules/dnn/src/model.cpp [hunk 5]
```diff
@@ -343,35 +359,45 @@ void DetectionModel::detect(InputArray frame, CV_OUT std::vector<int>& classIds,
...
+        {
+            std::map<int, std::vector<size_t> > class2indices;
+            for (size_t i = 0; i < predClassIds.size(); i++)
+            {
+                if (predConf[i] >= confThreshold)
+                {
+                    class2indices[predClassIds[i]].push_back(i);
+                }
+            }
+            for (auto it = class2indices.begin(); it != class2indices.end(); ++it)
```
Comment (dkurt) [inline modules/dnn/src/model.cpp:373]:
```cpp
for (const auto& it : class2indices)
```

File: modules/dnn/src/model.cpp [hunk 6]
```diff
@@ -343,35 +359,45 @@ void DetectionModel::detect(InputArray frame, CV_OUT std::vector<int>& classIds,
...
+                    localBoxes.push_back(predBoxes[idx]);
+                    localConfidences.push_back(predConf[idx]);
+                }
+                std::vector<int> indices;
+                NMSBoxes(localBoxes, localConfidences, confThreshold, nmsThreshold, indices);
+                for (int idx : indices)
+                {
+                    boxes.push_back(localBoxes[idx]);
+                    confidences.push_back(localConfidences[idx]);
+                    classIds.push_back(it->first);
```
Comment (dkurt) [inline modules/dnn/src/model.cpp:388]:
```cpp
classIds.resize(classIds.size() + indices.size(), it->first)
```

File: modules/dnn/include/opencv2/dnn/dnn.hpp (hunk 4 above)
Comment (l-bat) [reply modules/dnn/include/opencv2/dnn/dnn.hpp:1283]: Do you have any ideas?

Comment (dkurt) [reply modules/dnn/include/opencv2/dnn/dnn.hpp:1283]: Due this is just a procedure we can just define a static method in the `model.cpp` like
```cpp
void disableNMS(Net& net)
{
    for (String& name : net.getUnconnectedOutLayersNames())
    {
        int layerId = net.getLayerId(name);
        Ptr<RegionLayer> layer = net.getLayer(layerId).dynamicCast<RegionLayer>();
        if (!layer.empty())
        {
            layer->nmsThreshold = 0;
        }
    }
}
```

**Comment by dkurt** [comment]
Please check if failed test is caused by changes in this PR: https://pullrequest.opencv.org/buildbot/builders/precommit_custom_linux/builds/3884

**Comment by dkurt** [comment]
Observed some warnings on Windows. Looks like G-API related: https://pullrequest.opencv.org/buildbot/builders/precommit_custom_windows/builds/1067

<details>
<summary></summary>

```
c:\program files (x86)\intelswtools\openvino_2020.2.117\deployment_tools\inference_engine\include\ie_iextension.h(283): warning C4100: 'resp': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\src\backends\ie\giebackend.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_gapi.vcxproj]
c:\program files (x86)\intelswtools\openvino_2020.2.117\deployment_tools\inference_engine\include\ie_iextension.h(282): warning C4100: 'cnnLayer': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\src\backends\ie\giebackend.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_gapi.vcxproj]
c:\program files (x86)\intelswtools\openvino_2020.2.117\deployment_tools\inference_engine\include\ie_iextension.h(282): warning C4100: 'factory': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\src\backends\ie\giebackend.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_gapi.vcxproj]
c:\program files (x86)\intelswtools\openvino_2020.2.117\deployment_tools\inference_engine\include\ie_iextension.h(298): warning C4100: 'resp': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\src\backends\ie\giebackend.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_gapi.vcxproj]
c:\program files (x86)\intelswtools\openvino_2020.2.117\deployment_tools\inference_engine\include\ie_iextension.h(298): warning C4100: 'size': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\src\backends\ie\giebackend.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_gapi.vcxproj]
c:\program files (x86)\intelswtools\openvino_2020.2.117\deployment_tools\inference_engine\include\ie_iextension.h(298): warning C4100: 'types': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\src\backends\ie\giebackend.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_gapi.vcxproj]
c:\program files (x86)\intelswtools\openvino_2020.2.117\deployment_tools\inference_engine\include\ie_iextension.h(324): warning C4100: 'node': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\src\backends\ie\giebackend.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_gapi.vcxproj]
c:\program files (x86)\intelswtools\openvino_2020.2.117\deployment_tools\inference_engine\include\ie_iextension.h(334): warning C4100: 'implType': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\src\backends\ie\giebackend.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_gapi.vcxproj]
c:\program files (x86)\intelswtools\openvino_2020.2.117\deployment_tools\inference_engine\include\ie_iextension.h(334): warning C4100: 'node': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\src\backends\ie\giebackend.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_gapi.vcxproj]
C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\memory(913): warning C4996: 'InferenceEngine::ShapeInferExtension::~ShapeInferExtension': Use a common Extension interface (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\src\backends\ie\giebackend.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_gapi.vcxproj]
C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\memory(901): warning C4996: 'InferenceEngine::ShapeInferExtension::ShapeInferExtension': Use a common Extension interface (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\src\backends\ie\giebackend.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_gapi.vcxproj]
c:\program files (x86)\intelswtools\openvino_2020.2.117\deployment_tools\inference_engine\include\ie_iextension.h(283): warning C4100: 'resp': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\test\infer\gapi_infer_ie_test.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_test_gapi.vcxproj]
c:\program files (x86)\intelswtools\openvino_2020.2.117\deployment_tools\inference_engine\include\ie_iextension.h(282): warning C4100: 'cnnLayer': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\test\infer\gapi_infer_ie_test.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_test_gapi.vcxproj]
c:\program files (x86)\intelswtools\openvino_2020.2.117\deployment_tools\inference_engine\include\ie_iextension.h(282): warning C4100: 'factory': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\test\infer\gapi_infer_ie_test.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_test_gapi.vcxproj]
c:\program files (x86)\intelswtools\openvino_2020.2.117\deployment_tools\inference_engine\include\ie_iextension.h(298): warning C4100: 'resp': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\test\infer\gapi_infer_ie_test.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_test_gapi.vcxproj]
c:\program files (x86)\intelswtools\openvino_2020.2.117\deployment_tools\inference_engine\include\ie_iextension.h(298): warning C4100: 'size': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\test\infer\gapi_infer_ie_test.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_test_gapi.vcxproj]
c:\program files (x86)\intelswtools\openvino_2020.2.117\deployment_tools\inference_engine\include\ie_iextension.h(298): warning C4100: 'types': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\test\infer\gapi_infer_ie_test.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_test_gapi.vcxproj]
c:\program files (x86)\intelswtools\openvino_2020.2.117\deployment_tools\inference_engine\include\ie_iextension.h(324): warning C4100: 'node': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\test\infer\gapi_infer_ie_test.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_test_gapi.vcxproj]
c:\program files (x86)\intelswtools\openvino_2020.2.117\deployment_tools\inference_engine\include\ie_iextension.h(334): warning C4100: 'implType': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\test\infer\gapi_infer_ie_test.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_test_gapi.vcxproj]
c:\program files (x86)\intelswtools\openvino_2020.2.117\deployment_tools\inference_engine\include\ie_iextension.h(334): warning C4100: 'node': unreferenced formal parameter (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\test\infer\gapi_infer_ie_test.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_test_gapi.vcxproj]
C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\memory(913): warning C4996: 'InferenceEngine::ShapeInferExtension::~ShapeInferExtension': Use a common Extension interface (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\test\infer\gapi_infer_ie_test.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_test_gapi.vcxproj]
C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\memory(901): warning C4996: 'InferenceEngine::ShapeInferExtension::ShapeInferExtension': Use a common Extension interface (compiling source file C:\build\precommit_custom_windows\opencv\modules\gapi\test\infer\gapi_infer_ie_test.cpp) [C:\build\precommit_custom_windows\build\modules\gapi\opencv_test_gapi.vcxproj]
```

<details>

**Comment by dmatveev** [comment]
@dkurt those warnings are coming from IE, aren't they?

We didn't move to its new API though.

**Comment by dkurt** [comment]
@dmatveev, you're right, sorry for that.

**Pull request merged**